keps/sig-cli/4764-kubectl-watchlist/README.md

```diff
@@ -0,0 +1,767 @@
+
+<<[UNRESOLVED Will **watchlist** replace **list** completely]>>
+**watchlist** was first released in version 1.27 and is still in the alpha stage, with continuous improvements 
+over the subsequent releases. For a feature in such an early stage, it is too soon to say whether we will fully
+rely on it without other fallback mechanisms.
+
+In my opinion, due to the performance improvements, we should use **watchlist** by default instead of **list** in the
+long term. For the evolution strategy, we can follow the approach of `client-go` and address it in a later release.
+
+<<[/UNRESOLVED]>>
+
+
+### User Stories (Optional)
+
+#### Story 1
+
+Given that there is a large collection in the control plane, a user wants to fetch that collection using `kubectl get`.
+
+As a user, I shouldn't need to worry about the underlying details. I should simply update the version of `kubectl`
+to gain the performance improvements for free.
+
+#### Story 2
+
+As a cluster administrator, after upgrading the cluster, I should observe a significant reduction in the requests
+penetrating into etcd on the monitoring dashboard. Ideally, the majority of the requests should be cached by the 
+apiserver, rather than penetrating into etcd.
+
+### Risks and Mitigations
+
+For users, there are no visible changes to the user interface. However, since different APIs are called, 
+some users with restricted permissions may encounter authorization issues. Specifically, users might need
+to be granted additional **watch** permissions for the corresponding resources to ensure this mechanism 
+works properly. Fortunately, we have a fallback mechanism in this place. Even if users do not have the 
+additional permission, the old behavior should still work properly. In the worst case, some extra API 
+calls might cause a slight delay.
+
+Regarding security, users who already have **list** permissions inherently have access to all objects of this type.
+Therefore, I don't believe this change will introduce any additional security issues.
+
+## Design Details
+
+Since support for watch lists was merged in version 1.27, the API server allows clients to include the 
+`sendInitialEvents` query parameter when performing **watch** operations. This parameter instructs the 
+kube-apiserver to first send the existing members of the collection as `ADDED` events before pushing subsequent 
+objects changed events. A special `BOOKMARK` event with annotation `k8s.io/initial-events-end=true` will be sent 
+at the end to indicate the completion of the initial object push.
+
+This mechanism has been integrated into `client-go` in 1.31, and users who use typed or dynamic client can get
+this for free by setting env `KUBE_FEATURE_WatchListClient=true`. But unfortunately, in current implementation, 
+`kubectl` doesn't not use dynamic or typed client to list resource, it use raw `rest` client. So the auto switching
```

Comment:
could `kubectl` actually use `client-go` instead of the raw rest client ?

Reply:
I wish it could, but in my understanding, there is no easy way to do so. `kubectl` uses resource names instead of kinds to fetch resources. Even if it's possible, it will still be a big change.